README

                  ApMon - Application Monitoring API for Perl
                                 version 2.2.22
                  ===========================================
                                  November 2007
                         California Institute of Technology
                    =======================================

INTRODUCTION
============

ApMon is an API that can be used by any application to send monitoring
information to MonALISA services (http://monalisa.caltech.edu). The
monitoring data is sent as UDP datagrams to one or more hosts running MonALISA.
The MonALISA host may require a password enclosed in each datagram, for
authentication purposes. ApMon can also send datagrams that contain monitoring
information regarding the system or the application.


WHAT'S NEW IN VERSION 2.x
=========================

ApMon 2.x was extended with xApMon, a feature that allows it to send to the
MonALISA service datagrams with monitoring information regarding the system and
the application. This extension can be used completely on Linux systems and
partially on Windows systems.

In 2.2.0 protocol has changed (requires ML Service >= 2.4.10), each UDP packet 
includes 2 new INT32 fields to identify the sender and the sequence number. This 
will enable ML Service to detect lost packets, if it is configured to do so. 
(See the two new fields: instance_id and sequence_number).


INSTALLATION
============

The ApMon archive contains the following files in the ApMon module:
   - ApMon.pm - main ApMon module. It can be instantiated by users to send data.
   - ApMon/Common.pm - contains common functions for all other modules.
   - ApMon/XDRUtils.pm - contains functions that encode different values in XDR format
   - ApMon/ProcInfo.pm - procedures to monitor the system and a given application
   - ApMon/ConfigLoader.pm - manages configuration retrieval from multiple places
   - ApMon/BgMonitor.pm - handles the background monitoring of system and applications
   - README - this file
   - example/* - a set of examples with the usage of the ApMon module.
   - example/destinations.conf - a sample destinations file, for url/file configuration
   - MAN - a short description and API functions

To install this module type the following:
   perl Makefile.PL
   make
   make test
   make install


DEPENDENCIES
============

This module requires these other modules and libraries:
   Data::Dumper
   LWP::UserAgent
   Socket
   Exporter


USING APMON
===========

The ApMon.pm modules defines the ApMon class that has to be instantiated in one of the
following ways:

  - passing as parameters a list of destination locations. These are strings that
    represent either a filename, or an url address (if they start with http://). ApMon
    class will start a second process that will verify periodically these destination
    locations for any changes. This verification is done in a sepparate process and the
    settings are communicated to the main procress through a local file, stored in temp.
    This is because there could be long waiting times for the downloading and the
    resolving of the destination hosts for the udp datagrams. This way, ApMon sendParam*
    functions will not block.

  - passing as parameter a reference to an array, containg directly the destinations
    (host[:port][ pass]). ApMon will send datagrams to all valid given destinations (i.e.
    hosts that can be resolved), with default options. In this case, the child process
    that verifies configuration files for changes will be not be started and all function
    calls refering to this part will generate an error message.

  - passing as parameter to the constructor a reference to a hash. In this case, the keys
    will be destinations and the corresponding values will be references to another hash
    in which the key is a configuration option and the value is the value of that option.
    Note that in this case, the options should not be preceded by xApMon_ and options
    should be 0/1 instead of on/off as in the configuration file.
    In this case, the child process that verifies configuration files for changes will be
    not be started and all function calls refering to this part will generate an error
    message. If system and job monitoring and general information are disabled, the child
    process that performs this monitoring will not be starting. Any attempts to
    setMonitorClusterNode or setJobPID will generate errors.

  - passing no parameters. In this case, in order to initialize the destionations for the
    packets, you will have to use the setDestinations() function. This accepts the same
    parameters as the constructor, as described above.

  - passing as parameter the number 0. In this case no fork will be done and the user will
    have to initialize the destinations using the setDestinations() function. Also any
    background monitoring will be sent only when user invokes the sendBgMonitoring() routine.

  For a practical example, please see the EXAMPLE test script.

  The datagram sent to the MonaLisa module has the following structure:
  - a header which has the following syntax:
      v:<ApMon_version>p:<password>

      (the password is sent in plaintext; if the MonALISA host does not require
      a password, an 0-length string should be sent instead of the password).

  - instance_id ((int) from v2.2.0)
  - sequence_nr ((int) from v2.2.0) - these two are used to identify the ApMon sender
  - cluster name (string) - the name of the monitored cluster
  - node name (string) - the name of the monitored nodes
  - number of parameters (int)
  - for each parameter:
        - name (string),
        - value type (int),
        - value (can be double, int or string)
  - optionally a time (int) if the user wants to send a result specifying the time himself.

  The configuration file has the following format:

# list of destination MonALISA services, one on each line:
IP_address|DNS_name[:port][ password]

# list of xApMon_... settings, one on each line:
xApMon_option = on/off/value

  For an exmple see the destinations.conf file.

  If the port is not specified, the default value 8884 will be assumed.
If the password is not specified, an empty string will be sent as password
to the MonALISA host (and the host will only accept the datagram if it does not
require a password). The configuration file may contain blank lines and
comment lines (starting with "#"); these lines are ignored, and so are the
leading and the trailing white spaces from the other lines.

  To send user parameters to MonALISA, you have the following set of functions:

  - sendParameters ( $clusterName, $nodeName, @params);

Use this to send a set of parameters to all given destinations.
The default cluster an node names will be updated with the values given here.
If afterwards you want to send more parameters, you can use the shorter version
of this function, sendParams. The parameters to be sent can be eiter a list, or
a reference to a list, a hash or a reference to a hash with pairs. This list
should have an even length and should contain pairs like (paramName, paramValue).
paramValue can be a string, an integer or a float. Due to the way that Perl
interprets functions parameters, you can put as many parameters in the function
call as you want, not needing to create a list for this.

  - sendParams (@params);

Use this to send a set of parameters without specifying a cluster and a node name.
In this case, the default values for cluster and node name will be used. See the
sendParameters function for more details.

  - sendTimedParameters ($clusterName, $nodeName, $time, @params);

Use this instead of sendParameters to set the time for each packet that is sent.
The time is in seconds from Epoch. If you use the other function, the time for these
parameters will be sent by the MonALISA serice that receives them. Note that it is
recommended to use the other version unless you really need to send the parameters
with a different time, since the local time on the machine might not be synchronized
to a time server. The MonALISA service sets the correct real time for the packets
that it receives.

   - sendTimedParams ($time, @params);

This is the short version of the sendTimedParameters that uses the default cluster
and node name to sent the parameters and allows you to specify a time (in seconds
from Epoch) for each packet.

   Please see the EXAMPLE file for examples of using these functions.

  The configuration file and/or URLs can be periodically checked for changes,
but this option is disabled by default. In order to enable it, the user should
call setConfRecheck($bool, [$interval]); If you want, you can specify the time interval
at which the recheck operatins are performed.

  To monitor jobs, you have to specify the PID of the parent process for the tree of
processes that you want to monitor, the working directory, the cluster and the
node names. If work directory is "", no information will be retrieved about disk:
  - addJobToMonitor ($pid, $workDir, $clusterName, $nodeName);

  To stop monitoring a job, just call:
  - removeJobToMonitor ($pid);

  If you want to disable temporarily sending of background monitoring information, and
to enable it afterwards, you can use:
  - enableBgMonitoring ($bool)

  If you don't want to have the two background processes, you can turn them off whenever
you want using the following function:
  - stopBgProcesses ();

  In this case, configuration changes will not be notified anymore, and no background
monitoring will be performed. To force a configuration check or send monitoring information
about the system and/or jobs, you can use the following two functions:
  - refreshConfig ();
  - sendBgMonitoring ();

  To change the log-level of the ApMon module, you can either set it in the configuration file
or use the following function:
  - setLogLevel(level);
with the following valid log-levels: "DEBUG", "NOTICE", "INFO", "WARNING", "ERROR", "FATAL"

  To get in your program the monitored parameters, you can use the getSysMonInfo and 
getJobMonInfo specifying the (job's PID and) the parameters you want.

  To set the maxim number of messages that can be sent to a MonALISA service, per second,
you can use the follwing function:
  - setMaxMsgRate(rate);
By default, it is 50. This is a very large number, and the idea is to prevent errors from
the user. One can easily put in a for loop, without any sleep, some sendParams calls that
can generate a lot of unnecessary network load.

  If you have to recreate the ApMon object over and over again, then you should use the
following function when you fininshed using a instance of the ApMon:
  - free();
This will delete the temporary file and terminate the background processes. The UDP socket 
will not be closed, but if you will reuse ApMon afterwards, it will not open another socket.

  ATTENTION:
From version 2.2.0 a new mecanism was added to detect lost packets (see the two new fields:
instance_id and sequence_number). However, if you use free, and then recreate the ApMon instance,
these numbers will be also reinitialized and the new ApMon instance will be considered as
another, completly different sender. Thus, the usage of free is not encouraged.

  LIMITATIONS:
  The maximum size of a datagram is specified by the constant MAX_DGRAM_SIZE;
by default it is 8192B and the user should not modify this value as some hosts
may not support UDP datagrams larger than 8KB.


xApMon - MONITORING INFORMATION
===============================

ApMon can be configured to send to the MonALISA service monitoring information
about the application or about the system. This information is obtained
from the proc/ file system.

There are three categories of monitoring datagrams that ApMon can send:

a) job monitoring information - contains the following parameters:
    job_run_time        - elapsed time from the start of this job
    job_cpu_time        - processor time spent running this job
    job_cpu_usage       - percent of the processor used for this job, as reported by ps
    job_virtualmem      - virtual memory occupied by the job
    job_rss             - resident image size of the job
    job_mem_usage       - percent of the memory occupied by the job, as reported by ps
    job_workdir_size    - size in MB of the working directory of the job
    job_disk_total      - size in MB of the total size of the disk partition containing the working directory
    job_disk_used       - size in MB of the used disk partition containing the working directory
    job_disk_free       - size in MB of the free disk partition containing the working directory
    job_disk_usage      - percent of the used disk partition containing the working directory
    job_open_files	- number of open file descriptiors for this job


  To enable job monitoring put in the destination file the following lines:
    xApMon_job_monitoring = on
    xApMon_job_interval = 20
  To disable sending of the virtualmem parameter, put in the destination file:
    xApMon_job_virtualmem = off

b) system monitoring information - contains the following parameters:
    cpu_usr             - cpu-usage information
    cpu_sys             - all these will produce coresponding paramas without "sys_"
    cpu_nice
    cpu_idle
    cpu_iowait
    cpu_irq
    cpu_softirq
    cpu_steal
    cpu_guest
    cpu_usage		- in percent, (all - idle)
    interrupts		- the number of interrupts received (in interrupts/s)
    context_switches	- the number of context switches (in switches/s)
    load1               - system load information
    load5
    load15
    mem_used            - memory usage information
    mem_free

Changes
Revision history for Perl extension ApMon.

2.2.25
    - More flexible method to get the network interface names

2.2.24 Apr 2016
    - Accept any physical network interface name
    - Publish the total traffic per node (sum of all physical interfaces) as "total_traffic_{in,out}"
    - Publish the global IPv6 addresses (one per physical interface)
    - Compatibility with CentOS 7's ifconfig format (to extract IPv4 and IPv6 addresses)

2.2.23 Sep 2015
    - Add support for p4p* network interfaces


2.2.22 Nov 2014
    - Support em* interface names alongside legacy eth* names

2.2.21 Oct 2013
    - Fix parsing of /proc/net/dev that now has spaces after the interface name

2.2.20 Oct 2009
    - Support Solaris for host monitoring

2.2.19 Thu Jun 19 17:45:12 CEST 2008
    - Do not factor in CPU frequency fluctuations when reporting ksi2k values
    - Recognize TMPDIR as the default path where to put the temporary files instead of hardcoding /tmp/

2.2.18 Tue Feb 12 17:26:35 CET 2008
    - Enconding numbers larger than 2^31 as REAL instead of INT to preserve the actual value

2.2.17 Tue Nov 27 13:32:29 CET 2007
    - Fixed security issue: /tmp files were world writeable. Now 0600.

2.2.16 Tue Nov  6 15:43:20 CET 2007
    - Sending again the swap_used/free/usage parameters (broken since 2.2.11)
    - Avoid collecting BgMonitoring info immediately after initialization

2.2.15 Thu Nov  1 15:08:22 CET 2007
    - Finer-grained disk usage statistics (use df -k instead of df -m)
    - Simpler overflow check

2.2.14 Thu Oct  4 11:40:34 CEST 2007
    - If known, report also the ksi2k factor for the system (as generic info)

2.2.13 Mon Sep  3 10:23:40 CEST 2007
    - Reporting the size and success of the sent datagram
    - Fix for handling string params in sendToML.sh
    - Requiring LWP::Useragent only when needed, so if config is not on http, this is not required

2.2.12 Fri Aug 24 16:33:33 CEST 2007
    - Support changing generic values
    - Support variable CpuMhz when reporting consumed ksi2k
    - Adding reporting for minor and major faults of the processes of a monitored job

2.2.11 Mon Jul  2 18:05:43 CEST 2007
    - Adding swap_(in|out)_R also for kernel >2.5
    - Renaming pages_(in|out) to blocks_(in|out)_R and adding it also for kernel >2.5
    - adding cpu_iowait, interrupts and context_switches
    - adding sendToML.sh and servMon.sh scripts to the distribution

2.2.10 Tue Jun  5 18:16:00 CEST 2007
    - Fixed support for cpu type and parameters on ia64.
    - adding /etc/issue as a place to get distribution name

2.2.9 Tue Feb  6 10:44:48 CET 2007
    - fixed /tmp not accesible problem: sending both user and monitoring information was not
      working. Now, /tmp fails while running, old config is kept. In order to avoid all this,
      create ApMon(0) to avoid using bg processes.

2.2.8 Fri Sep  1 16:26:17 CEST 2006
    - added 2 functions: getSysMonInfo & getJobMonInfo to get the last values of all monitored 
      parameters from ApMon.

2.2.7 Mon Jul 31 16:05:39 CEST 2006
    - don't print an error if no config url is given (but there is a destination)
    - another fix in parsing uptime

2.2.6 Thu Jul  6 21:22:31 CEST 2006
    - small fix in handling errors when loading config from multiple urls
    - added a new example: muliple_urls.pl to show how to init from several urls

2.2.5 Mon Jun 26 12:01:41 CEST 2006
    - fix in parsing uptime
    - cpu_time is always increasing, even if one of the forked processes finishes
    - added run_ksi2k to report the elapsed time (not CPU time) in ksi2k units.

2.2.4 Tue Jun 13 11:54:50 CEST 2006
    - added mem_actualfree: actually free memory: free + buffers + cached

2.2.3 Tue May 30 19:12:04 CEST 2006
    - added cpu_cache - as in /proc/cpuinfo
    - added possibility to report cpu_ksi2k as cpu consumed power, per second, in Kilo SpecINT2000 units.

2.2.2 Wed May 24 22:43:52 CEST 2006
    - changed the Instance ID to be more random
    - added new parameters: kernel_version, platform, os_type, mem_buffers, mem_cached
    
2.2.1 Tue Mar 7 10:12:34 CET 2006
    - changed ps -ax -e to ps -A to avoid the warning on debian and to keep the consistency on Mac.

2.2.0 Sun Feb 19 21:52:52 CET 2006
    - Changed protocol (requires ML Service >= 2.4.10) so that each UDP packet includes 2 new INT32 fields
      to identify the sender and the sequence number. This will enable ML Service to detect lost packets,
      if it is configured to do so.

2.1.1 Mon Feb 13 23:07:39 CET 2006
    - sendBgMonitoring accepts now a parameter to force sending the bgMonitoring
      regardles of how much time passed since last sent
    - LoadAvg (load1, load5, load15) works on Darwin (Mac OS X)
    - Process monitoring works on Darwin (Mac OS X)
    - fixed reporting cummulative values (ethX_yyy, swap in/out, pages in/out, cpu_zzz)
    
2.0.10 Wed Dec  7 17:04:26 CET 2005
    - don't compute the used_swap and used_mem unless we have values for total and free mem and swap
    - added sockets info (tcp, udp, unix) and connection states for tcp sockets from netstat
    - sorted the list of sent parameters - it should be easier to find them in the GUI client
    - decreased the default max rate of sent messages to 20 per second
    
2.0.9 Sat Nov  5 21:45:41 CET 2005
    - don't try to encode parameters with undefined values

2.0.8 Thu Oct 27 21:49:57 CEST 2005
    - fixed div by zero if there is no swap space
    - produce the number of running, sleeping, io blocked, zombie, stopped processes

2.0.7 Mon Oct  3 22:25:58 CEST 2005
    - replaced `hostname -f` with Net::Domain::getfqdn() for Mac compatibility
    
2.0.6 Fri Sep  2 16:33:59 CEST 2005
    - added a limit of the maximum number of messages that are sent, per second
    - added a function to free the resources when you have to recreate the object
    - fixed a bug in computing the ethX_in/out network traffic when system counters were wrapping 

2.0.5 Sat Aug  6 18:40:45 CEST 2005

ApMon/Common.pm
package ApMon::Common;

use strict;
use warnings;

require Exporter;
use Carp qw(cluck);
use Socket;
use ApMon::XDRUtils;
use Data::Dumper;
use Sys::Hostname;

use vars qw(@ISA @EXPORT @EXPORT_OK $APMON_DEFAULT_PORT $VERSION %defaultOptions $KSI2K $CpuMHz);

push @ISA, qw(Exporter);
push @EXPORT, qw(logger);
push @EXPORT_OK, qw($APMON_DEFAULT_PORT %defaultOptions);

$VERSION = "2.2.18";
$APMON_DEFAULT_PORT = 8884;

my @LOG_LEVELS = ("DEBUG", "NOTICE", "INFO", "WARNING", "ERROR", "FATAL");
my $CRT_LOGLEVEL = 2;	# index in the array above

my $MAX_MSG_RATE = 20;	# Default value for max nr. of messages that user is allowed to send, per second

$KSI2K = undef;		# kilo spec ints 2k for this machine

$CpuMHz = undef;	# Cpu Speed when taking the speed for KSI2k

# Default options for background monitoring
%defaultOptions = (
        'job_monitoring' => 1,          # perform (or not) job monitoring
        'job_interval' => 60,           # at this interval (in seconds)
        'job_data_sent' => 0,           # time from Epoch when job information was sent; don't touch!

        'job_cpu_time' => 1,            # processor time spent running this job in seconds
	'job_cpu_ksi2k' => 1,		# used CPU power in ksi2k units (see SpecInt2000 for details);
        'job_run_time' => 1,            # elapsed time from the start of this job in seconds
	'job_run_ksi2k' => 1,		# elapsed time in ksi2k units
        'job_cpu_usage' => 1,           # current percent of the processor used for this job, as reported by ps
        'job_virtualmem' => 1,          # size in JB of the virtual memory occupied by the job, as reported by ps
        'job_rss' => 1,                 # size in KB of the resident image size of the job, as reported by ps
        'job_mem_usage' => 1,           # percent of the memory occupied by the job, as reported by ps
        'job_workdir_size' => 1,        # size in MB of the working directory of the job
        'job_disk_total' => 1,          # size in MB of the total size of the disk partition containing the working directory
        'job_disk_used' => 1,           # size in MB of the used disk partition containing the working directory
        'job_disk_free' => 1,           # size in MB of the free disk partition containing the working directory
        'job_disk_usage' => 1,          # percent of the used disk partition containing the working directory
	'job_open_files' => 1,		# number of open file descriptors
	'job_page_faults_min' => 1,	# number of minor page faults in the job
	'job_page_faults_maj' => 1,	# number of major page faults in the job


        'sys_monitoring' => 1,          # perform (or not) system monitoring
        'sys_interval' => 60,           # at this interval (in seconds)
        'sys_data_sent' => 0,           # time from Epoch when system information was sent; don't touch!

        'sys_cpu_usr' => 1,             # cpu-usage information
        'sys_cpu_sys' => 1,             # all these will produce coresponding paramas without "sys_"
        'sys_cpu_nice' => 1,
        'sys_cpu_idle' => 1,
	'sys_cpu_iowait' => 1,
	'sys_cpu_irq' => 1,
	'sys_cpu_softirq' => 1,
	'sys_cpu_steal' => 1,
	'sys_cpu_guest' => 1,
        'sys_cpu_usage' => 1,
	'sys_interrupts' => 1,
	'sys_context_switches' => 1,
        'sys_load1' => 1,               # system load information
        'sys_load5' => 1,
        'sys_load15' => 1,
        'sys_mem_used' => 1,            # memory usage information
        'sys_mem_free' => 1,
	'sys_mem_actualfree' => 1,	# actually free memory: free + cached + buffers
        'sys_mem_usage' => 1,
	'sys_mem_buffers' => 1,
	'sys_mem_cached' => 1,
        'sys_blocks_in' => 1,
        'sys_blocks_out' => 1,
        'sys_swap_used' => 1,           # swap usage information
        'sys_swap_free' => 1,
        'sys_swap_usage' => 1,
        'sys_swap_in' => 1,
        'sys_swap_out' => 1,
        'sys_net_in' => 1,              # network transfer in kBps
        'sys_net_out' => 1,             # these will produce params called ethX_in, ethX_out, ethX_errs
        'sys_net_errs' => 1,            # for each eth interface
	'sys_net_sockets' => 1,		# number of opened sockets for each proto => sockets_tcp/udp/unix ...
	'sys_net_tcp_details' => 1,	# number of tcp sockets in each state => sockets_tcp_LISTEN, ...
        'sys_processes' => 1,		# total processes and processs in each state (R, S, D ...)
	'sys_uptime' => 1,		# uptime of the machine, in days (float number)


        'general_info' => 1,            # send (or not) general host information once every 2 $sys_interval seconds
        'general_data_sent' => 0,       # time from Epoch when general information was sent; don't touch!

        'hostname' => 1,
        'ip' => 1,                      # will produce ethX_ip params for each interface
	'kernel_version' => 1,
	'platform' => 1,
	'os_type' => 1,
        'cpu_MHz' => 1,
        'no_CPUs' => 1,                 # number of CPUs
	'ksi2k_factor' => 1,		# system's ksi2k factor, if known
        'total_mem' => 1,
        'total_swap' => 1,
	'cpu_vendor_id' => 1,
	'cpu_family' => 1,
	'cpu_model' => 1,
	'cpu_model_name' => 1,
	'cpu_cache' => 1,
	'bogomips' => 1);


# Create a UDP socket through which all information is sent
if(! socket(SOCKET, PF_INET, SOCK_DGRAM, getprotobyname("udp"))){
        logger("FATAL", "Cannot create UDP socket $@");
	die;
}

# Simple logger
sub logger {
        my ($level, $msg) = @_;
        my $i = 0;
        $i++ while (! ($LOG_LEVELS[$i] eq $level) and ($i < @LOG_LEVELS));
        if($CRT_LOGLEVEL <= $i and $i < @LOG_LEVELS){
		my $now =localtime();
		$now =~ s/^\S+\s((\S+\s+){3}).*$/$1/;
                print $now."ApMon[$LOG_LEVELS[$i]]: $msg\n";
        }
}

# Sets the CRT_LOGLEVEL
sub setLogLevel {
	my $level = shift;
	logger("NOTICE", "Setting loglevel to $level");
	if(! defined $level){
		cluck("got undefined level from");
		return;
	}
	my $i = 0;
	$i++ while (! ($LOG_LEVELS[$i] eq $level) and ($i < @LOG_LEVELS));
	if($i < @LOG_LEVELS){
		$CRT_LOGLEVEL = $i;
	}else{
		logger("WARNING", "Unknown log level \"$level\" - ignoring.\n");
	}
}

# Sets the maximum rate for sending messages (see shouldSend subroutine)
sub setMaxMsgRate {
	my $rate = shift;
	
	$MAX_MSG_RATE = $rate;
	logger("INFO", "Setting maxMsgRate to $rate");
}

# For each destination, we'll keep a pair (instance_id, seq_nr) that will identify us
my $senderRef = {};
my $instance_id = getInstanceID();

# This is used internally to send a set of parameters to a given destination.
sub directSendParameters {
        my ($destination, $clusterName, $nodeName, $time, $paramsRef) = @_;

        my @params;
	if(! defined($paramsRef)){
		logger("WARNING", "Not sending undefined parameters!");
		return;
	}
	if(! defined($time)){
		logger("WARNING", "Not sending the parameters for an undefined time!");
		return;
	}

	if(! shouldSend()){
		#logger("WARNING", "Not sending since the messages are too often!");
		return;
	}
	
        if(ref($paramsRef->[0]) eq "ARRAY"){
                @params = @{$paramsRef->[0]};
        }elsif(ref($paramsRef->[0]) eq "HASH"){
                @params = %{$paramsRef->[0]};
        }else{
                @params = @$paramsRef;
        }

        if(@params == 0){
                return;
        }
	
	$senderRef->{$destination} = {INSTANCE_ID => $instance_id, SEQ_NR => 0} if ! $senderRef->{$destination};
	my $sender = $senderRef->{$destination};
	$sender->{INSTANCE_ID} = ($$ << 16) | ($sender->{INSTANCE_ID} && 0xffff);
	$sender->{SEQ_NR} = ($sender->{SEQ_NR} + 1) % 2_000_000_000; # wrap around 2 mld
	
        my ($host, $port, $pass) = split(/:/, $destination);
        logger("NOTICE", "====> $host|$port|$pass/$clusterName/$nodeName".($time != -1 ? " @ $time" : "")." [$sender->{SEQ_NR} # $sender->{INSTANCE_ID}]");
        for(my $i = 0; $i < @params; $i += 2){
		if(defined($params[$i]) && defined($params[$i+1])){
	                logger("NOTICE", "  ==> $params[$i] = $params[$i+1]");
		}else{
			logger("NOTICE", "  ==> ".(defined($params[$i]) ? $params[$i] : "undef name")." = ".(defined($params[$i+1]) ? $params[$i+1] : "undef value")." <== ignoring pair");
			splice(@params, $i, 2);
			$i-=2;
		}
        }
        my $header = "v:${VERSION}_plp:$pass";
        my $msg = ApMon::XDRUtils::encodeString($header)
		. ApMon::XDRUtils::encodeINT32($sender->{INSTANCE_ID})
		. ApMon::XDRUtils::encodeINT32($sender->{SEQ_NR})
		. ApMon::XDRUtils::encodeParameters($clusterName, $nodeName, $time, @params);
        my $in_addr = inet_aton($host);
        my $in_paddr = sockaddr_in($port, $in_addr);
	my $msg_len = length($msg);
        if(send(SOCKET, $msg, 0, $in_paddr) != $msg_len){
                logger("ERROR", "Could not send UDP datagram to $host:$port");
        }else{
		logger("NOTICE", "~~~~> Packet sent successfully; total size=$msg_len bytes.");
	}
}

# This is called by child processes to read messages (if they exist) from the parent.
sub readMessage {
        my $PIPE = shift;

        my ($rin, $win, $ein, $rout, $wout, $eout) = ('', '', '');
        my $retMsg = "";
        vec($rin,fileno($PIPE),1) = 1;
        $ein = $rin | $win;
        my ($nfound,$timeleft) = select($rout=$rin, $wout=$win, $eout=$ein, 0);
        if($nfound){
                sysread($PIPE, $retMsg, 1024);
                logger("DEBUG", "readMessage: $retMsg");
	}
	return $retMsg;
}

# This is called by main process to send a message to a child that reads form the given pipe
sub writeMessage {
	my ($PIPE, $msg) = @_;

	if(defined $PIPE){
		logger("DEBUG", "writeMessage: $msg");
                syswrite($PIPE, $msg);
        }else{
                logger("ERROR", "Trying to send '$msg' to child, but the pipe is not defined!");
        }
}

# copy the time when last data was sent
sub updateLastSentTime {
	my ($srcOpts, $dstOpts) = @_;

	$dstOpts->{'general_data_sent'} = $srcOpts->{'general_data_sent'} if $srcOpts->{'general_data_sent'};
	$dstOpts->{'sys_data_sent'} = $srcOpts->{'sys_data_sent'} if $srcOpts->{'sys_data_sent'};
	$dstOpts->{'job_data_sent'} = $srcOpts->{'job_data_sent'} if $srcOpts->{'job_data_sent'};
}

# This is used to update the configuration for an object that has in it's base hash the following elements
# DESTINATIONS, CONF_RECHECK, LAST_CONF_CHECK_TIME, CONF_CHECK_INTERVAL and CONF_FILE.
# In practice, both ApMon and BgMonitor use it to update their configuration.
sub updateConfig {
        my $this = shift;

	if(! $this->{ALLOW_BG_PROCESSES}){
		$this->{DESTINATIONS} = $this->{CONFIG_LOADER}->{DESTINATIONS};
		return;
	}
        my $now = time;
        if((scalar(keys %{$this->{DESTINATIONS}}) > 0 and $this->{CONF_RECHECK} == 0)
                        or ($this->{LAST_CONF_CHECK_TIME} + $this->{CONF_CHECK_INTERVAL} > $now)){
                return;
        }
        logger("DEBUG", "Updating configuration from $this->{CONF_FILE}");
        if(open(CONF, "<$this->{CONF_FILE}")){
		my $prevDest = $this->{DESTINATIONS} || {};
                $this->{DESTINATIONS} = {};     # clear old destinations first
                my ($crtDest, $line);
                while($line = <CONF>){
                        chomp $line;
                        if($line =~ /^(\S+):(\S+):(\S*)$/){
                                # reading a new destination
                                $crtDest = $line;
                                my %defOpts = %defaultOptions;  #get a copy of the default options
                                $this->{DESTINATIONS}->{$crtDest}->{OPTS} = \%defOpts;
				updateLastSentTime($prevDest->{$crtDest}->{OPTS}, $this->{DESTINATIONS}->{$crtDest}->{OPTS});
				$this->{DESTINATIONS}->{$crtDest}->{PREV_RAW_DATA} = 
					($prevDest->{$crtDest}->{PREV_RAW_DATA} ? $prevDest->{$crtDest}->{PREV_RAW_DATA} : {});
                                logger("DEBUG", "Adding destination $line");
                        }elsif($line =~ /^\s(\S+)=(\S+)/) {
                                # reading an attribute for the current destination and modify the current options
                                my ($name, $value) = ($1, $2);
                                logger("DEBUG", "Adding $name=$value");
				if($name eq 'loglevel'){
					$this->setLogLevel($value);
                                }elsif($name eq 'conf_recheck'){
                                        $this->{CONF_RECHECK} = $value;
                                }elsif($name eq 'recheck_interval'){
                                        $this->{CONF_CHECK_INTERVAL} = $value;
				}elsif($name eq 'maxMsgRate'){
                                        $this->setMaxMsgRate($value);
				}else{
                                        $this->{DESTINATIONS}->{$crtDest}->{OPTS}->{$name} = $value;
                                }
                        }else{
                                logger("WARNING", "Unknown line in conf file: $line");
                        }
                }
                close CONF;
        }else{
                logger("ERROR", "Error opening temporary config file $this->{CONF_FILE}. Current config is unchanged.");
                return;
        }
        $this->{LAST_CONF_CHECK_TIME} = time;
}

# don't allow a user to send more than MAX_MSG messages per second, in average
my $prvTime = 0;
my $prvSent = 0;
my $prvDrop = 0;
my $crtTime = 0;
my $crtSent = 0;
my $crtDrop = 0;
my $hWeight = 0.92;

# Decide if the current datagram should be sent.
# This decision is based on the number of messages previously sent.
sub shouldSend {
	my $now = time;

	if($now != $crtTime){
		# new time
		# update previous counters;
		$prvSent = $hWeight * $prvSent + (1 - $hWeight) * $crtSent / ($now - $crtTime); 
		$prvTime = $crtTime;
		logger("DEBUG", "previously sent: $crtSent; dropped: $crtDrop");
		# reset current counter
		$crtTime = $now;
		$crtSent = 0;
		$crtDrop = 0;
	}
	my $valSent = $prvSent * $hWeight + $crtSent * (1 - $hWeight); # compute the history
	
	my $doSend = 1;
	my $level = $MAX_MSG_RATE - $MAX_MSG_RATE / 10; # when we should start dropping messages
	
	if($valSent > $MAX_MSG_RATE - $level){
		$doSend = rand($MAX_MSG_RATE / 10) < ($MAX_MSG_RATE - $valSent);
	}
	
	# counting sent and dropped messages
	if($doSend){
		$crtSent++;
	}else{
		$crtDrop++;
	}
	
	return $doSend;
}

# Try to generate a more random instance id. It takes the process ID and
# combines it with the last digit from the IP addess and a random number
sub getInstanceID {
	my $pid = $$;
	my $ip = int(rand(256));  # last digit of the ip address
	my $host = hostname();    # from Sys::Hostname
	if($host){
		my $addr = inet_ntoa(scalar gethostbyname($host));
		$ip = $1 if $addr =~ /(\d+)$/;
	}
	my $rnd = int(rand(256));
	my $iid = ($pid << 16) | ($ip << 8) | $rnd; # from all this, generate the instance id
	return $iid;
}

# Try to determine the CPU type. Returns a hash with: cpu_model_name, cpu_MHz, cpu_cache (in KB)
# TODO: make this work also for Mac.
sub getCpuType {
	my $cpu_type = {};
	if(-r "/proc/cpuinfo"){
		if(open(CPU_INFO, "</proc/cpuinfo")){
			my $line;
			while($line = <CPU_INFO>){
				if($line =~ /cpu MHz\s+:\s+(\d+\.?\d*)/){
					$cpu_type->{"cpu_MHz"} = $1;
					$CpuMHz = $1;
				}
				$cpu_type->{"cpu_model_name"} = $1 if($line =~ /model name\s+:\s+(.+)/ || $line =~ /family\s+:\s+(.+)/);
				$cpu_type->{"cpu_cache"} = $1 if($line =~ /cache size\s+:\s+(\d+)/);
			}
			close(CPU_INFO);
		}else{
			logger("NOTICE", "Cannot open /proc/cpuinfo");
		}
	}
	if(-r "/proc/pal/cpu0/cache_info"){
		if(open(CACHE_INFO, "</proc/pal/cpu0/cache_info")){
			my $line;
			my $level3params = 0;
			while($line = <CACHE_INFO>){
				$level3params = 1 if($line =~/Cache level 3/);
				$cpu_type->{"cpu_cache"} = $1 / 1024 if ($level3params && $line =~ /Size\s+:\s+(\d+)/);
			}
			close(CACHE_INFO);
		}else{
			logger("NOTICE", "Cannot open /proc/pal/cpu0/cache_info");
		}
	}
	if(! scalar(keys(%$cpu_type))){
		logger("INFO", "Cannot get cpu type");
		return undef;
	}
	return $cpu_type;
}

# Set the SI2K performance meter for this machine. If this function is called then parameter
# cpu_ksi2k will also be reported for the job monitoring with a value computed this way:
# cpu_ksi2k(job) = cpu_time(job) * ( si2k / 1000)
sub setCpuSI2k {
	my $si2k = shift;
	$KSI2K = $si2k / 1000.0 if($si2k);
}

1;

